### PR TITLE
raft cf memtable insert with hint prefix extractor

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
raft cf memtable insert with hint prefix extractor.
@siddontang @BusyJay PTAL

### CODE REVIEW

File: src/raftstore/store/keys.rs [hunk 1]
```diff
@@ -80,6 +80,10 @@ pub fn region_raft_prefix(region_id: u64) -> Vec<u8> {
     key
 }
 
+pub fn region_raft_prefix_len() -> usize {
+    REGION_RAFT_PREFIX_KEY.len() + mem::size_of::<u64>()
```
Comment (siddontang) [inline src/raftstore/store/keys.rs]: do we need to handle apply raft key, raft state key and raft log key separately?  They have same raft prefix but with different postfix:

```
pub const RAFT_LOG_SUFFIX: u8 = 0x01;
pub const RAFT_STATE_SUFFIX: u8 = 0x02;
pub const APPLY_STATE_SUFFIX: u8 = 0x03;
```

Comment (siddontang) [inline src/raftstore/store/keys.rs]: but seem that if we can update raft log first, we can find raft state with the hint directly.

**Comment by siddontang** [comment]
any performance increased ?

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -404,7 +405,11 @@ fn get_rocksdb_write_cf_option(config: &toml::Value) -> RocksdbOptions {
 }
 
 fn get_rocksdb_raftlog_cf_option(config: &toml::Value) -> RocksdbOptions {
-    get_rocksdb_cf_option(config, "raftcf", 256 * 1024 * 1024, false, false)
+    let mut opt = get_rocksdb_cf_option(config, "raftcf", 256 * 1024 * 1024, false, false);
+    opt.set_memtable_insert_hint_prefix_extractor("RaftPrefixSliceTransform",
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: use a config to do benchmark easily.

**Comment by zhangjinpeng1987** [comment]
I bench this branch and branch master several times, this pr increase insert qps by 15%-17%, and reduce avg latency by 15%.

this branch
```
1st time
read/write requests:                 1600000 (4455.55 per sec.)
avg:                                            7.32ms
approx.  99 percentile:             13.22ms

2nd time
read/write requests:                 1600000 (4368.64 per sec.)
avg:                                            7.18ms
approx.  99 percentile:             15.55ms
```
master
```
1st time:
read/write requests:                 1600000 (3735.74 per sec.)
avg:                                            8.56ms
approx.  99 percentile:             15.75ms
2nd time:
read/write requests:                 1600000 (3806.85 per sec.)
avg:                                            8.40ms
approx.  99 percentile:             16.29ms
```

**Comment by zhangjinpeng1987** [comment]
ref https://github.com/pingcap/tikv/issues/1603

**Comment by ngaut** [comment]
Is the bench result created by sysbench ?

**Comment by zhangjinpeng1987** [comment]
Yes, the result is created by sysbench. 
When bench random insert, both this branch and master's bottle neck is the cpu of raftstore thread, from the result we get a primary conclusion that this branch reduce about 15% cpu usage of raftstore thread in heavy write case.

File: etc/config-template.toml [hunk 3]
```diff
@@ -233,6 +233,9 @@ max-bytes-for-level-base = "128MB"
 target-file-size-base = "32MB"
 block-cache-size = "256MB"
 
+# memtable insert with hint prefix extractor.
+# memtable-insert-with-hint = true
```
Comment (ngaut) [inline etc/config-template.toml]: Why add a new config, any side effective ?

Comment (siddontang) [inline etc/config-template.toml]: just for benchmark before, can be removed now.

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

**Pull request merged**